streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala

```diff
@@ -45,11 +45,8 @@ class KGroupedStream[K, V](val inner: KGroupedStreamJ[K, V]) {
    * represent the latest (rolling) count (i.e., number of records) for each key
    * @see `org.apache.kafka.streams.kstream.KGroupedStream#count`
    */
-  def count()(implicit materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] = {
-    val c: KTable[K, java.lang.Long] =
-      inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, ByteArrayKeyValueStore]])
-    c.mapValues[Long](Long2long _)
-  }
+  def count()(implicit materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] =
+    inner.aggregate((() => 0L).asInitializer, ((_: K, _: V, agg: Long) => agg + 1L).asAggregator, materialized)
```

Comment:
Why this change? Is that intentional?

Reply:
Ditto below.

Reply:
It is. Without this change, my unit my unit tests fail with this exception:

```
Caused by: java.lang.ClassCastException: java.lang.Long cannot be cast to [B
	at org.apache.kafka.common.serialization.ByteArraySerializer.serialize(ByteArraySerializer.java:19)
	at org.apache.kafka.streams.kstream.internals.FullChangeSerde$1.serialize(FullChangeSerde.java:73)
	at org.apache.kafka.streams.kstream.internals.FullChangeSerde$1.serialize(FullChangeSerde.java:60)
	at org.apache.kafka.streams.kstream.internals.suppress.KTableSuppressProcessor.buffer(KTableSuppressProcessor.java:96)
	at org.apache.kafka.streams.kstream.internals.suppress.KTableSuppressProcessor.process(KTableSuppressProcessor.java:87)
	at org.apache.kafka.streams.kstream.internals.suppress.KTableSuppressProcessor.process(KTableSuppressProcessor.java:40)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:117)
```

It looks like this is because the value serde is not passed down here: 
* https://github.com/apache/kafka/blob/4824dc994d7fc56b7540b643a78aadb4bdd0f14d/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedTable.scala#L50
* https://github.com/apache/kafka/blob/4824dc994d7fc56b7540b643a78aadb4bdd0f14d/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala#L51

The `KTable` returned by `mapValues` has a null value serde which causes it to fallback to the configured default (byte array serde in this case), which is incorrect.

If you have other ideas for how to fix this or prefer that I put it in a different PR (since it is sort of orthogonal as you said), let me know!

Reply:
cc @vvcephei could you take a look?

Reply:
That's right @casey-green . As I see it, the two options are:
1. implement a scala-native count (which you've done)
2. propagate the serdes (see below)

Propagating the serdes would look like this:
```scala
    val tableImpl = javaCountTable.asInstanceOf[KTableImpl[K, ByteArrayKeyValueStore, lang.Long]]
    javaCountTable.mapValues[Long](
      (l => long2Long(l)) : ValueMapper[lang.Long, Long],
      Materialized.`with`[K, Long, ByteArrayKeyValueStore](tableImpl.keySerde(), Serdes.Long)
    )
```
... pretty nasty.

To be honest, I was never a fan of the extra `mapValues` node there. It seemed wasteful to me. I think in a vacuum, I prefer your solution.

However, dropping a node from the topology results in compatibility problems. If we were to elide the `mapValues` now, then people would have to perform an application reset when upgrading to this version. We'd probably better avoid that.

Thoughts?

Reply:
Ah, good point about compatibility! In this case, I think the somewhat nastier code is preferable over forcing people to reset their applications after upgrading because we removed the `mapValues` operator.

I went ahead and made these changes in each of the three `count` methods.